### PR TITLE
Main

### PR DESCRIPTION
**Pull request by Ritchie-R-H**
Integrate vgg to wasserstein.py; add unit test "wasserstein_test.py"

### CODE REVIEW

File: codex/loss/wasserstein.py [hunk 1]
```diff
@@ -124,6 +125,9 @@ def wasserstein_distortion(
     raise ValueError(
         f"features and `log2_sigma` must have same spatial shape, but received "
         f"{features_a.shape[-2:]} and {log2_sigma.shape}, respectively.")
+  if jnp.max(log2_sigma) > num_levels: # add an available check
```
Comment (jonaballe) [inline codex/loss/wasserstein.py]: This might not work if the function is compiled with `jax.jit`. Can you add a unit test that checks that the function can be compiled?

File: codex/loss/features.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  selected_keys = ["relu1_2", "relu2_2", "relu3_3", "relu4_3", "relu5_3"]
+
+  features = [out[key].squeeze(0) for key in selected_keys]
+
+  # for idx, f in enumerate(features):
+  #     print(f"Feature {selected_keys[idx]} shape: {f.shape}")
+
+  return features
+
+if __name__ == '__main__':
```
Comment (jonaballe) [inline codex/loss/features.py]: Please remove this section. This will go into the library, so unnecessary parts of the code should be removed.

File: codex/loss/wasserstein_test.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    assert jnp.allclose(ls, expected_ls, atol=0.1), "Sigma not scaled correctly"
+    return jnp.zeros(())
+
+    # Monkeypatch wasserstein_distortion to capture log2_sigma
+    with pytest.MonkeyPatch.context() as mp:
+      mp.setattr("wasserstein.wasserstein_distortion", check_sigma)
+      multi_wasserstein_distortion(features, features, log2_sigma)
+
+
+if __name__ == "__main__":
```
Comment (jonaballe) [inline codex/loss/wasserstein_test.py]: Please remove this section. Tests should be run using pytest.

File: codex/loss/wasserstein.py [hunk 4]
```diff
@@ -227,3 +231,22 @@ def multi_wasserstein_distortion(
...
+  features_a = vgg16(img1)
+  features_b = vgg16(img2)
+
+  log2_sigma = jnp.full((224, 64), 4) # TODO Test different values of sigma
+  result = multi_wasserstein_distortion(features_a, features_b, log2_sigma)
+
+  return result
+
+
+if __name__ == "__main__":
```
Comment (jonaballe) [inline codex/loss/wasserstein.py]: Please remove this section. This will go into the library, so unnecessary parts of the code should be removed.

File: codex/loss/wasserstein.py [hunk 5]
```diff
@@ -24,6 +24,7 @@
 import jax
 from jax import numpy as jnp
 from codex.ops import gradient
+from features import vgg16
```
Comment (jonaballe) [inline codex/loss/wasserstein.py]: For imports in the library, please use the same format as above, e.g. in this case:

```
from codex.loss import features
```

and then below, access `features.vgg16` instead of `vgg16`.

File: codex/loss/wasserstein_test.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+  # Adjusted sigma should be 4 - log2(64/32) = 3.0
+  expected_ls = jnp.full((32, 32), 3.0)
+
+  def check_sigma(fa, fb, ls, **kwargs):
+    assert jnp.allclose(ls, expected_ls, atol=0.1), "Sigma not scaled correctly"
+    return jnp.zeros(())
+
+    # Monkeypatch wasserstein_distortion to capture log2_sigma
+    with pytest.MonkeyPatch.context() as mp:
```
Comment (jonaballe) [inline codex/loss/wasserstein_test.py]: This code cannot be reached. Please look at the "Problems" section in VSCode. If your environment is set up correctly, it will show sanity checks like this. For example, line too long, or code unreachable, etc.

File: codex/loss/wasserstein_test.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
+import jax
+import jax.numpy as jnp
+import numpy as np
+import pytest
+from wasserstein import lowpass, compute_multiscale_stats, wasserstein_distortion, \
```
Comment (jonaballe) [inline codex/loss/wasserstein_test.py]: For library code, please use the format
```
from codex.loss import wasserstein
```
and then call `wasserstein.lowpass` and so on below.

File: codex/loss/features.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import jax
+import jax.numpy as jnp
+import flaxmodels as fm
+
+def vgg16(image):
+  """
+  Extract specific layer's feature maps of an image passing VGG
+  """
+  # Load image
+  img = Image.open(image)
```
Comment (jonaballe) [inline codex/loss/features.py]: Let's keep the loading of the image and of the VGG model separate from the computing of the vgg features. This is because in a training loop, we might only want to load the VGG model and the images once, but we might want to compute the features many times.

File: .gitignore [hunk 9]
```diff
@@ -23,3 +23,6 @@ poetry.lock
 
 # PyCharm
 .idea
+
```
Comment (jonaballe) [inline .gitignore]: You can keep it in your own git fork, but let's not add it to the library in general.

**Review by jonaballe — CHANGES_REQUESTED** [review]
